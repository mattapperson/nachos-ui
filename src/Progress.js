import React from 'react'
import PropTypes from 'prop-types'
import { View } from 'react-native'

const defaultTheme = {
  PROGRESS_BACKGROUND: '#bdc1cc',
  PROGRESS_ACTIVE_COLOR: '#2f8cff',
  PROGRESS_HEIGHT: 6,
  PROGRESS_WIDTH: 300,
}

const Progress = (props) => {
  const {
    theme,
    progress,
    style,
    color,
    width = theme.PROGRESS_WIDTH,
    height = theme.PROGRESS_HEIGHT,
  } = props

  const outerStyle = {
    borderRadius: height / 2,
    ...style,
    width,
    height,
  }

  const innerStyle = {
    height,
    width: width * progress,
    backgroundColor: color || undefined,
    borderRadius: height / 2,
  }

  return (
    <View style={[theme.outerStyle, outerStyle]}>
      <View style={[theme.innerStyle, innerStyle]} />
    </View>
  )
}

Progress.defaultStyle = (theme = defaultTheme) => {
  return {
    outerStyle: {
      backgroundColor: theme.PROGRESS_BACKGROUND,
      overflow: 'hidden',
    },
    innerStyle: {
      backgroundColor: theme.PROGRESS_ACTIVE_COLOR,
    },
  }
}

Progress.propTypes = {
  progress: PropTypes.number.isRequired,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  width: PropTypes.number,
  height: PropTypes.number,
  color: PropTypes.string,
  theme: PropTypes.object,
}

Progress.defaultProps = {
  progress: 0,
}

export default Progress
