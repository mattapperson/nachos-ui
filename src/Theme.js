import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
