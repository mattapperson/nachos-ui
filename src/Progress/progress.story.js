import { storiesOf } from "../storybook";
import { withInfo } from "../storybook";
import React from "react";
import Progress from "./index";
import { View } from "react-native";

const Example = () => {
  const progressStyle = { margin: 15 };

  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Progress progress={0.1} style={progressStyle} />
      <Progress progress={0.5} style={progressStyle} />
    </View>
  );
};

storiesOf("Progress", module).add("default", withInfo()(Example));

export default Example;
