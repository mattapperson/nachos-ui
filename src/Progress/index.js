import React from "react";
import PropTypes from "prop-types";
import { View } from "react-native";
import { withTheme } from "../Theme";

const Progress = props => {
  const { theme, progress, style, color, width, height } = props;

  const outerStyle = {
    borderRadius: height / 2,
    width,
    height
  };

  const innerStyle = {
    height,
    width: width * progress,
    borderRadius: height / 2
  };
  if (color) innerStyle.backgroundColor = color;

  return (
    <View style={[theme.outerStyle, outerStyle, style]}>
      <View style={[theme.innerStyle, innerStyle]} />
    </View>
  );
};

Progress.themeConfig = {
  props: {
    width: 300,
    height: 6
  },
  style: {
    outerStyle: {
      backgroundColor: "@disabledColor",
      overflow: "hidden"
    },
    innerStyle: {
      backgroundColor: "@primaryColor"
    }
  }
};

Progress.propTypes = {
  progress: PropTypes.number.isRequired,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  width: PropTypes.number,
  height: PropTypes.number,
  color: PropTypes.string,
  theme: PropTypes.object
};

Progress.defaultProps = {
  progress: 0
};

export default withTheme("Progress", Progress);
