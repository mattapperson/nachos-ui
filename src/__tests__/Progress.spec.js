import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "../Theme";

import Progress from "../Progress";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Progress progress={0.5} />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
