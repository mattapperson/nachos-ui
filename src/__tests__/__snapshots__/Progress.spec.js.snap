// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-backgroundColor-xtitrs rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-overflowX-11yh6sk rn-overflowY-buy8e9 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  style={
    Object {
      "borderBottomLeftRadius": "3px",
      "borderBottomRightRadius": "3px",
      "borderTopLeftRadius": "3px",
      "borderTopRightRadius": "3px",
      "height": "6px",
      "width": "300px",
    }
  }
>
  <div
    className="rn-alignItems-1oszu61 rn-backgroundColor-1otax7f rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
    style={
      Object {
        "borderBottomLeftRadius": "3px",
        "borderBottomRightRadius": "3px",
        "borderTopLeftRadius": "3px",
        "borderTopRightRadius": "3px",
        "height": "6px",
        "width": "150px",
      }
    }
  />
</div>
`;
